8509 Household Context report optimization for 2026 SPM and APR

ttoomey opened the pull request:
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch

### Description
Implements a pre-computed household context layer for FY2026 HUD reports to centralize business logic and improve generation performance.

### Functional Areas
- **Household Context Layer**: Introduced `HudReports::HouseholdContext` model and database table to cache pre-calculated household attributes (chronic status, move-in dates, household composition) per report run.
- **Business Logic Centralization**: Extracted core household rules (chronic inheritance, household typing, move-in date logic) into `HudReports::HouseholdLogic`, removing duplication between runtime concerns and the new pre-computation layer.
- **FY2026 Report Generators**: Refactored APR, CAPER, CE APR, and DQ generators to populate household contexts during the `prepare_report` phase and utilize `AprClientBuilder` for efficient snapshot creation.
- **SPM Optimization**: Configured SPM HDX upload to generate household contexts once and share them with subordinate Data Quality reports via the `source_report_id` mechanism.
- **Legacy Compatibility**: Updated `HudReports::Households` concern to delegate calculation logic to the new `HouseholdLogic` class, ensuring consistency across older report versions.
- **Maintenance**: Added a daily maintenance task to prune expired household context records.

### Type of Change
Refactor

**TBD:**
- Allow sharing households between SPM and DQ reports to avoid redundant builds.
- Periodically prune stale `HudReports::HouseholdContext` records.
- Expand builder pattern to other reports currently relying on the `HudReports::Households` module.

### Type of Change
Refactor

Review complete: 14 files analyzed (3 high, 6 medium, 5 low risk)
Findings: 0 critical, 4 warning, 3 suggestion

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [ ] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] I have updated the documentation (or not applicable)
- [x] I have added spec tests (or not applicable)
- [ ] I have provided testing instructions in this PR or the related issue (or not applicable)

ttoomey committed:
wip household contexts

ttoomey committed:
wip

ttoomey committed:
Merge branch 'release-193' into 8509-materialized-households

ttoomey committed:
fixes

ttoomey committed:
Merge branch 'release-193' into 8509-materialized-households

ttoomey committed:
rubocop

ttoomey committed:
spec

ttoomey committed:
Merge branch 'release-193' into 8509-materialized-households

ttoomey committed:
wip

ttoomey committed:
Merge branch 'release-193' into 8509-materialized-households

ttoomey committed:
add compatability proxy

ttoomey committed:
fix lints and failures

ttoomey committed:
refinement

ttoomey committed:
refinements & fixes

ttoomey committed:
bugfix

nil household_members on hud_report_apr_clients

ttoomey committed:
handle missing hoh

ttoomey committed:
rubocop

ttoomey committed:
fix

ttoomey committed:
fix ambiguous col

ttoomey committed:
fix hoh_exit_date was missing in the generator class

ttoomey committed:
optimization

ttoomey committed:
fixes

ttoomey committed:
Merge branch 'release-193' into 8509-materialized-households

ttoomey committed:
fixes

ttoomey committed:
Fixed ambiguous SQL select, strict leavers

ttoomey committed:
precompute veteran flags, refactor veterans concern

ttoomey committed:
docs

ttoomey committed:
cleanup

ttoomey committed:
testkit docs

ttoomey committed:
fix

ttoomey committed:
bugfix

ttoomey committed:
fixes

ttoomey committed:
fix sql table mapping with dynamic override

ttoomey committed:
wip

ttoomey committed:
docs

ttoomey committed:
fix hh_ctx with col ref

ttoomey committed:
revert docs

ttoomey committed:
reduce scope creep, target apr client build only

ttoomey committed:
unify household logic

ttoomey committed:
Merge branch 'release-194' into 8509-materialized-households

ttoomey committed:
cleanups

ttoomey committed:
wip

ttoomey committed:
apr builder refactor

ttoomey committed:
Merge branch 'release-194' into 8509-materialized-households

ttoomey committed:
household members

ttoomey committed:
wip

ttoomey committed:
rm q3 spec

ttoomey committed:
fix crash

ttoomey committed:
cleanup api

ttoomey committed:
wip

ttoomey committed:
support shared report household context

ttoomey committed:
refinements

ttoomey commented:
design artifact

# Pre-computed Household Context for HUD Reports

This design shifts complex household-level calculations—chronic status inheritance, move-in date derivation, and population categorization—from runtime Ruby calculations into a pre-computed database layer.

## Design Goals

**Performance**: Compute complex household attributes once per report run rather than repeatedly during snapshot generation.

**Consistency**: Centralize business rules in a single source of truth, ensuring identical behavior across all report types and fiscal years.

**Maintainability**: Replace scattered runtime logic with a clear separation between computation (preparation phase) and consumption (snapshot generation phase).

## Architecture

### Core Components

- **`HudReports::HouseholdContext`**: Database table storing pre-computed household attributes per enrollment. Acts as a logic-specific cache layer between raw warehouse data and report snapshots.

- **`HudReports::HouseholdLogic`**: Single source of truth for all household business rules. Defines chronic status inheritance, move-in date propagation, household typing, and parenting youth identification.

- **`HudReports::HouseholdContextBuilder`**: Orchestrates the pre-computation process. Batches enrollments by household, applies `HouseholdLogic`, and bulk-imports results.

- **`AprClientBuilder`**: Maps pre-computed context fields and enrollment data into final report snapshot attributes.

### Two-Phase Execution

**Preparation Phase**: `HouseholdContextBuilder` analyzes all in-scope enrollments, applies business rules, and persists results to the context table.

**Snapshot Phase**: Report generators load contexts in batches and construct snapshot records using pre-computed values, eliminating expensive joins and repeated calculations.

## Data Model

The `hud_report_household_contexts` table caches:

- **Inherited attributes**: Chronic status, move-in date, household type
- **Head of Household snapshots**: Entry date, move-in date, CoC code, age, veteran status
- **Household aggregates**: Member count, age ranges, veteran flags, parenting youth indicators

This eliminates the need to traverse household relationships during snapshot generation.

## Cross-Report Optimization

For multi-report processes (e.g., SPM HDX Upload generating multiple DQ sub-reports):

1. **Shared Build**: Parent report computes contexts for all projects in scope
2. **Efficient Sharing**: Child reports copy only the subset of contexts they need via direct database operations
3. **Validation**: System enforces matching date ranges between source and target reports

This reduces redundant computation by 60-80% for complex report workflows.

## Backward Compatibility

Pre-FY2026 reports continue using runtime calculations via the `HudReports::Households` concern. This concern was refactored to delegate to `HouseholdLogic`, ensuring older fiscal years use identical business rules as the new optimized path.

FY2026+ generators bypass the concern entirely, reading from the pre-computed context table instead.

## Verification

If report values appear incorrect, inspect the `hud_report_household_contexts` table for the enrollment in question. The cached values should reflect the applied business rules. Test coverage exists at both the logic layer (`HouseholdLogic`) and orchestration layer (`HouseholdContextBuilder`).

eanders commented on app/jobs/importing/run_daily_imports_job.rb:
```diff
@@ -177,6 +177,11 @@ def _perform
       end
 
+      run_maintenance_task('Prune HUD report data') do
+        HudReports::HouseholdContext.prune!
```
Haven't looked yet, but want to confirm this doesn't break any of the view logic for a report.  Assuming this is ephemeral and can safely be deleted, computation results have been stored on the relevant report client/enrollment records.

ttoomey replied:
Yes, the household contexts exist only to populate the apr client (or spm enrollment records soon). I tried to avoid them leaking out into the question layer; questions should be unchanged although it's possible something could leak through with the household module refactor; I'll do a re-review to check

eanders commented on app/models/hud_reports/household_context.rb:
```diff
@@ -0,0 +1,53 @@
+
+    def self.prune!
+      # Delete contexts for reports older than 2 weeks or reports that have been deleted
+      valid_report_ids = HudReports::ReportInstance.where('created_at >= ?', 2.weeks.ago).select(:id)
```
Maybe we want to keep letting simple queries be written as strings, but I'd like to try and keep things in standard active record syntax where possible.  I think this would generate the same query:

```suggestion
      valid_report_ids = HudReports::ReportInstance.where(created_at: ..2.weeks.ago).select(:id)
```

eanders commented on app/models/hud_reports/household_context_builder.rb:
```diff
@@ -0,0 +1,281 @@
+      she.household_id || "#{she.enrollment_group_id}*HH"
+    end
+
+    def each_household_id_batch
```
This is probably fine, as it may cast a wider than intended net, but it is probably worth noting that `HouseholdID` is only guaranteed  unique within a data source, so there's a chance you'll load and generate household information that is never used.  Again, probably safe, but might be worth a note.

eanders replied:
Or, maybe that's not true since you're starting from the SHE scoped to the included clients.

eanders commented on app/models/hud_reports/household_context_builder.rb:
```diff
@@ -0,0 +1,281 @@
+          pit_chronic_at_start = chronic_at_start
+        else
+          pit_chronic_at_start = enrollment&.chronically_homeless_at_start(date: report_date)
+        end
```
Is this because it is less expensive/we already need to use `chronically_homeless_at_start`?

```suggestion
        # If we happen to be looking for chronicity on the entry date, we don't need to recalculate it, save some processing and re-use this.
        if report_date.blank? || report_date == enrollment&.EntryDate
          pit_chronic_at_start = chronic_at_start
        else
          pit_chronic_at_start = enrollment&.chronically_homeless_at_start(date: report_date)
        end
```

ttoomey replied:
yes, it's a performance optimization, I'll add a comment

eanders commented on app/models/hud_reports/household_context_builder.rb:
```diff
@@ -0,0 +1,281 @@
+
+    def build_contexts_for_household(hh_id, members)
+      # Pre-calculate common HH values
+      hh_member_hashes = members.map do |m|
```
`members` are SHE, we might want to rename that.

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2026/apr_client_builder.rb:
```diff
@@ -0,0 +1,466 @@
+    # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+    def map_standard_attributes
+      exit_date = @last_service_history_enrollment.last_date_in_program
+      exit_record = @last_service_history_enrollment.enrollment if exit_date.present? && exit_date <= @report.end_date
```
This doesn't look right to me.  It's calling an Enrollment `exit_record`.  Maybe `exited_enrollment` would be better.

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2026/apr_client_builder.rb:
```diff
@@ -0,0 +1,466 @@
+      @a_t ||= HudApr::Fy2020::AprClient.arel_table
+    end
+
+    def initialize(report:, client:, enrollments:, context_map:, hoh_enrollment_map:, needs_ce_assessments:, households:)
```
Can you add some comments about what each of these "map" things are?  I noticed later it calls `@ctx.age`, but it is unclear to me if that's the correct age calculation since I'm not sure what `@ctx` holds.

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2026/base.rb:
```diff
@@ -65,369 +65,67 @@ def needs_ce_assessments?
+          next unless result[:success]
+
+          # Handle duplication check here (Orchestration concern)
+          if processed_source_clients.include?(result[:source_client_id])
```
What might cause this?

ttoomey replied:
Maybe too defensive, not expected during normal operation.  But could be:
* maybe due to splitting/merging destination client while report is active?
* we have a join product in client scope

Note that this is a carry-over from the legacy implementation, for example the 2023 APR generator has the same guard (drivers/hud_apr/app/models/hud_apr/generators/shared/fy2023/base.rb:142)

ttoomey committed:
Merge branch 'release-194' into 8509-materialized-households

ttoomey committed:
8509 Household Context optimization for SPM (#6146)

Refactor SPM reports to use household contexts; improving memory use and consistency; brakeman upgrade, data_source key fix

ttoomey committed:
adapt change in a5cdf8a8da1aebc674d11181ee9321db1762bead

ttoomey committed:
Merge branch 'release-195' into 8509-materialized-households

ttoomey committed:
Merge branch 'release-195' into 8509-materialized-households

ttoomey commented:
Deployed to QA for testing with larger data set in EKS env. SPM generated in about 2 hours for ~9,000 client records with 141,000 enrollments. Completed without running out of memory

https://qa-warehouse.openpath.host/hud_reports/spms/==3UjLXGKfeiqAyRYGclzLoQ==

ttoomey committed:
Merge branch 'release-195' into 8509-materialized-households

ttoomey committed:
spm m1b [ci-focus: drivers/hud_spm_report/spec/models/datalab_testkit/all_projects_spec.rb]

ttoomey committed:
debugging

ttoomey committed:
Merge branch 'release-195' into 8509-materialized-households

ttoomey committed:
testkit [ci-focus: spec/models/test_empty_job_spec.rb]

ttoomey committed:
debug [ci-focus: spec/models/test_empty_job_spec.rb]

ttoomey committed:
debug [ci-focus: spec/models/test_empty_job_spec.rb]

ttoomey committed:
testkit [ci-focus: drivers/hud_spm_report/spec/models/datalab_testkit/all_projects_spec.rb]

ttoomey committed:
wip

ttoomey committed:
testkit [ci-focus: drivers/hud_spm_report/spec/models/datalab_testkit/all_projects_spec.rb]

ttoomey committed:
m1 validation specs

ttoomey committed:
testkit [ci-focus: drivers/hud_spm_report/spec/models/datalab_testkit/all_projects_spec.rb]

Correct M1b TH criteria, optimize bed night lookback, and format LOT to 2 decimals
* Logic: Applied "Literally Homeless at Entry" criteria to Transitional Housing in Measure 1b.
* Performance: Added a 7-year lookback filter to historical bed night loading.
* Formatting: Rounded Average and Median LOT calculations to two decimal places.
* Tests: Updated validation and unit tests to reflect the new criteria and formatting.

ttoomey committed:
rubocop

ttoomey committed:
testkit [ci-focus: drivers/hud_spm_report/spec/models/datalab_testkit/all_projects_spec.rb]

literally homeless fix?
th stayer inclusion

ttoomey committed:
revert

ttoomey committed:
pr feedback

ttoomey committed:
Merge branch 'release-196' into 8509-materialized-households

ttoomey committed:
docs

ttoomey committed:
household fixes

ttoomey committed:
conditionally force testkit checks

ttoomey committed:
debugging [ci-focus: drivers/hud_spm_report/spec/models/datalab_testkit/all_projects_spec.rb] testkit-check-all-results

ttoomey committed:
Merge branch 'release-196' into 8509-materialized-households

ttoomey committed:
correct clobbered spm enrollment file

ttoomey committed:
debugging [ci-focus: drivers/hud_spm_report/spec/models/datalab_testkit/all_projects_spec.rb] testkit-check-all-results

ttoomey committed:
speculative fix for slight undercounts in testkit

ttoomey committed:
debugging [ci-focus: drivers/hud_spm_report/spec/models/datalab_testkit/all_projects_spec.rb] testkit-check-all-results

ttoomey committed:
stricter move in date

ttoomey committed:
debugging [ci-focus: drivers/hud_spm_report/spec/models/datalab_testkit/all_projects_spec.rb] testkit-check-all-results

ttoomey committed:
rubocop

ttoomey committed:
fix expired gpg key in CI?

ttoomey committed:
debugging [ci-focus: drivers/hud_spm_report/spec/models/datalab_testkit/all_projects_spec.rb] testkit-check-all-results

ttoomey committed:
Merge branch 'release-196' into 8509-materialized-households

ttoomey committed:
Revert "speculative fix for slight undercounts in testkit"

This reverts commit 64ca7217b6d2bfcaa03bf86c2b645b3b70fe9d5d.

ttoomey committed:
revert

ttoomey committed:
fixes & cleanups

ttoomey committed:
Merge branch 'release-196' into 8509-materialized-households

ttoomey committed:
household sort order

ttoomey committed:
docs

ttoomey committed:
fix issue

ttoomey committed:
Merge branch 'release-203' into 8509-materialized-households

ttoomey committed:
rubocop

ttoomey committed:
fix rubocop complexity lints in build_contexts_for_household

ttoomey committed:
improve houshold logic specs

ttoomey committed:
cleanups from code review

ttoomey committed:
ce apr regression fix

ttoomey committed:
rubocop

ttoomey committed:
fixes

ttoomey committed:
Merge branch 'release-203' into 8509-materialized-households

ttoomey committed:
fix

ttoomey committed:
cleanups

ttoomey committed:
docs

ttoomey committed:
tidy fields

ttoomey commented:
this PR has been split into phases for easier review:

* https://github.com/greenriver/hmis-warehouse/pull/6303
* https://github.com/greenriver/hmis-warehouse/pull/6304
* https://github.com/greenriver/hmis-warehouse/pull/6305 (you are here)
